File: libPyQ/states.py
from math import sqrt, sin, cos
import numpy as np
from su import Pauli
from mat_func import proj


def bell(j, k):
    if j == 0 and k == 0:
        return np.array([[1/sqrt(2)], [0], [0], [1/sqrt(2)]])  # phi+
    elif j == 0 and k == 1:
        return np.array([[0], [1/sqrt(2)], [1/sqrt(2)], [0]])  # psi+
    elif j == 1 and k == 0:
        return np.array([[1/sqrt(2)], [0], [0], [-1/sqrt(2)]])  # phi-
    elif j == 1 and k == 1:
        return np.array([[0], [1/sqrt(2)], [-1/sqrt(2)], [0]])  # psi-


def Werner(w):
    return ((1-w)/4)*np.eye(4) + w*proj(4, Bell(1, 1))


def rhoBD(c1, c2, c3):
    return (1/4)*(np.eye(4) + c1*np.kron(Pauli(1), Pauli(1))
                  + c2*np.kron(Pauli(2), Pauli(2))
                  + c3*np.kron(Pauli(3), Pauli(3)))


def cb(d, j):
    cbs = np.zeros(d)
    cbs[j] = 1
    return cbs


def psi1qb(theta, phi):
    return cos(theta/2.0)*cb(2, 0) + (cos(phi)
                                      + sin(phi)*1j)*sin(theta/2.0)*cb(2, 1)


def rho1qb(r1, r2, r3):
    return 0.5*(Pauli(0) + r1*Pauli(1) + r2*Pauli(2) + r3*Pauli(3))


def rho2qb(CM):
    return (1/4)*(CM[0][0]*np.kron(Pauli(0), Pauli(0))
                  + CM[1][0]*np.kron(Pauli(1), Pauli(0))
                  + CM[2][0]*np.kron(Pauli(2), Pauli(0))
                  + CM[3][0]*np.kron(Pauli(3), Pauli(0))
                  + CM[0][1]*np.kron(Pauli(0), Pauli(1))
                  + CM[1][1]*np.kron(Pauli(1), Pauli(1))
                  + CM[2][1]*np.kron(Pauli(2), Pauli(1))
                  + CM[3][1]*np.kron(Pauli(3), Pauli(1))
                  + CM[0][2]*np.kron(Pauli(0), Pauli(2))
                  + CM[1][2]*np.kron(Pauli(1), Pauli(2))
                  + CM[2][2]*np.kron(Pauli(2), Pauli(2))
                  + CM[3][2]*np.kron(Pauli(3), Pauli(2))
                  + CM[0][3]*np.kron(Pauli(0), Pauli(3))
                  + CM[1][3]*np.kron(Pauli(1), Pauli(3))
                  + CM[2][3]*np.kron(Pauli(2), Pauli(3))
                  + CM[3][3]*np.kron(Pauli(3), Pauli(3)))

File: LibCQ/states.c

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <complex.h>

void psi1qb(double *theta, double *phi, double _Complex *psi) {
  *(psi+0) = cos((*theta)/2.0);  
  *(psi+1) = sin((*theta)/2.0)*(cos((*phi)) + I*sin((*phi)));
}

void bell(double _Complex *phip, double _Complex *psip, double _Complex *phim, double _Complex *psim){
  double f = 1.0/sqrt(2.0);
  phip[0] = f; phip[1] = 0.0; phip[2] = 0.0; phip[3] = f;
  phim[0] = f; phim[1] = 0.0; phim[2] = 0.0; phim[3] = -f;
  psip[0] = 0.0; psip[1] = f; psip[2] = f; psip[3] = 0.0;
  psim[0] = 0.0; psim[1] = f; psim[2] = -f; psim[3] = 0.0;
}

void werner2qb(double *p, double _Complex rho[][4]){
  double _Complex I4[4][4] = {{1,0,0,0},{0,1,0,0},{0,0,1,0},{0,0,0,1}};
  double _Complex proj[4][4] = {{0,0,0,0},{0,0.5,-0.5,0},{0,-0.5,0.5,0},{0,0,0,0}};
  int j, k;
  for(j = 0; j < 4; j++){
    for(k = 0; k < 4; k++){
      rho[j][k] = (1.0-(*p))*0.25*I4[j][k] + (*p)*proj[j][k];
    }
  }
}
